lib/matplotlib/axes/_axes.py

```diff
@@ -6580,6 +6580,12 @@ def hist(self, x, bins=None, range=None, density=None, weights=None,
                 mlast[:] = m
             tops.append(m)
 
+        # align != 'mid' only makes sense for equal-sized bins.
+        if align != 'mid' and not cbook._is_equally_spaced(bins):
```

Comment:
I think this should only apply to the step-like functions, so `if histtype in ['step', 'stepfilled']` ...   
For the bar-like functions, it may not make too much sense, but at least it's not wrong.

Reply:
I think #12494 shows very clearly that whoever implemented `align` did not take non-equally-spaced bins into account, and while the result is "not wrong", it doesn't make much sense either.